pageserver/src/tenant.rs

```diff
@@ -1751,12 +1783,67 @@ impl Tenant {
         }
     }
 
+    /// Shutdown the tenant and join all of the spawned tasks.
+    ///
+    /// The method caters for all use-cases:
+    /// - pageserver shutdown (freeze_and_flush == true)
+    /// - detach + ignore (freeze_and_flush == false)
+    ///
+    /// This will attempt to shutdown even if tenant is broken.
+    pub(crate) async fn shutdown(&self, freeze_and_flush: bool) -> Result<(), ShutdownError> {
+        debug_assert_current_span_has_tenant_id();
+        // Set tenant (and its timlines) to Stoppping state.
+        //
+        // Since we can only transition into Stopping state after activation is complete,
+        // run it in a JoinSet so all tenants have a chance to stop before we get SIGKILLed.
+        //
+        // Transitioning tenants to Stopping state has a couple of non-obvious side effects:
+        // 1. Lock out any new requests to the tenants.
+        // 2. Signal cancellation to WAL receivers (we wait on it below).
+        // 3. Signal cancellation for other tenant background loops.
+        // 4. ???
+        //
+        // The waiting for the cancellation is not done uniformly.
+        // We certainly wait for WAL receivers to shut down.
+        // That is necessary so that no new data comes in before the freeze_and_flush.
+        // But the tenant background loops are joined-on in our caller.
+        // It's mesed up.
+        // we just ignore the failure to stop
+        match self.set_stopping().await {
+            Ok(()) => {}
+            Err(SetStoppingError::Broken) => {
+                // assume that this is acceptable
+            }
+            Err(SetStoppingError::AlreadyStopping) => return Err(ShutdownError::AlreadyStopping),
+        };
+
+        if freeze_and_flush {
+            // Shut down all existing walreceiver connections and stop making the new ones.
```

Comment:
> and stop making the new ones.

Not sure I understand how that works, is it because of the stopping state checked here https://github.com/zenithdb/zenith/blob/8909ec3cd1cac12ea512bcd711c7cb6b380cabc5/pageserver/src/tenant/timeline/walreceiver/connection_manager.rs#L54?

Reply:
If so I think it makes sense to place it above on self.set_stopping

Reply:
Yes the walreceiver should exit on `set_stopping`, agreed, perhaps this message should not have been spared.

Reply:
```suggestion
```

Reply:
Edited the comment in [28d328b](https://github.com/neondatabase/neon/pull/4407/commits/28d328b664ac4088b32fdf2dd43052bc8efed45f)

Reply:
Thanks!